---
**Importing Libraries:**
* To get started we will use Python for data pre-processing .
* Import python libraries as necessary to get started for data load and later import other libraries as needed
---

In [419]:
#Import all necessary libraries
import numpy as np 
import pandas as pd 


---
# Dataset Load
---

**Dataset:**

* **train_data.csv** - training data with multiple statement dates per customer_ID
* **train_labels.csv** - target label for each customer_ID


---

Let us check size of dataset CSV file

---
Considering large number of rows around 5.5 million in **train_data.csv** dataset, using nrows option to load first 200000 rows from dataset file for Model building.

---

Load train_data.csv dataset file using nrows=200000

In [420]:
# Loading dataset train_data.csv
df = pd.read_csv('../input/amex-default-prediction/train_data.csv', nrows=200000)

In [421]:
# get shape of dataframe
print('Shape of dataset is:', df.shape)

Shape of dataset is: (200000, 190)


In [422]:
# print summary of dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Columns: 190 entries, customer_ID to D_145
dtypes: float64(185), int64(1), object(4)
memory usage: 289.9+ MB


**Observations:**

* There are total 190 variables in train_data.csv dataset
    * There are 185 variables(Columns) as dtype float64, 1 variable(Column) as dtype int64 and 4 variables(Columns) as dtype object

---
Need to load **train_labels.csv** for customer_ID with target label as 1 for Default and 0 for Not Default

---

In [423]:
# Loading dataset train_labels.csv
df_label = pd.read_csv('../input/amex-default-prediction/train_labels.csv')

In [424]:
# get shape of dataframe
print('Shape of dataset is:', df_label.shape)

Shape of dataset is: (458913, 2)


In [425]:
# print summary of dataframe
df_label.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458913 entries, 0 to 458912
Data columns (total 2 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   customer_ID  458913 non-null  object
 1   target       458913 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 7.0+ MB


**Observations:**

* There are total 458,913 entries for target label with customer_ID
* There is variable (column) customer_ID which has dtype as object and variable (column) target which has dtype as int64

---
Need to merge train_labels dataset with train_data dataset for target label.

---

In [426]:
# Merge of train_df_sample and train_label_df dataframe using key as customer_ID
df = pd.merge(df, df_label, how="inner", on=["customer_ID"])

In [427]:
# print summary of merged dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 0 to 199999
Columns: 191 entries, customer_ID to target
dtypes: float64(185), int64(2), object(4)
memory usage: 293.0+ MB


In [428]:
#display first 10 rows using head
df.head(10)

customer_ID         S_2       P_2  \
0  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-03-09  0.938469   
1  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-04-07  0.936665   
2  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-05-28  0.954180   
3  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-06-13  0.960384   
4  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-07-16  0.947248   
5  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-08-04  0.945964   
6  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-09-18  0.940705   
7  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-10-08  0.914767   
8  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-11-20  0.950845   
9  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-12-04  0.868580   

       D_39       B_1       B_2       R_1       S_3      D_41       B_3  D_42  \
0  0.001733  0.008724  1.006838  0.009228  0.124035  0.008771  0.004709   NaN   
1  0.005775  0.004923  1.000653  0.006151  0.126750  0.000798  0.002714   NaN   
2  0.091505  0.021655  1.009672  0.006815  0.123977  0.007598  0.009423   NaN   
3  0.002455  0.013683  1.002700  0.001373  0.117169  0.000685  0.005531   NaN   
4  0.002483  0.015193  1.000727  0.007605  0.117325  0.004653  0.009312   NaN   
5  0.001746  0.007863  1.005006  0.004220  0.110946  0.009857  0.009866   NaN   
6  0.002183  0.018859  1.008024  0.004509  0.103329  0.006603  0.000783   NaN   
7  0.003029  0.014324  1.000242  0.000263  0.108115  0.009527  0.007836   NaN   
8  0.009896  0.016888  1.003995  0.001789  0.102792  0.002519  0.009817   NaN   
9  0.001082  0.001930  1.007504  0.001772  0.100470  0.004626  0.006073   NaN   

   D_43      D_44       B_4      D_45       B_5       R_2      D_46      D_47  \
0   NaN  0.000630  0.080986  0.708906  0.170600  0.006204  0.358587  0.525351   
1   NaN  0.002526  0.069419  0.712795  0.113239  0.006206  0.353630  0.521311   
2   NaN  0.007605  0.068839  0.720884  0.060492  0.003259  0.334650  0.524568   
3   NaN  0.006406  0.055630  0.723997  0.166782  0.009918  0.323271  0.530929   
4   NaN  0.007731  0.038862  0.720619  0.143630  0.006667  0.231009  0.529305   
5   NaN  0.004128  0.027265  0.721371  0.138303  0.006437  0.275963  0.529762   
6   NaN  0.009397  0.006608  0.721725  0.108522  0.002254  0.307869  0.535747   
7   NaN  0.001399  0.005677  0.723293  0.187638  0.008154  0.519619  0.533295   
8   NaN  0.002578  0.013713  0.730918  0.178574  0.009845  0.430077  0.539065   
9   NaN  0.007595  0.005182  0.735001  0.189011  0.001795  0.430871  0.542119   

       D_48  D_49       B_6       B_7       B_8      D_50      D_51       B_9  \
0  0.255736   NaN  0.063902  0.059416  0.006466  0.148698  1.335856  0.008207   
1  0.223329   NaN  0.065261  0.057744  0.001614  0.149723  1.339794  0.008373   
2  0.189424   NaN  0.066982  0.056647  0.005126  0.151955  1.337179  0.009355   
3  0.135586   NaN  0.083720  0.049253  0.001418  0.151219  1.339909  0.006782   
4       NaN   NaN  0.075900  0.048918  0.001199  0.154026  1.341735  0.000519   
5       NaN   NaN  0.095784  0.035738  0.004577  0.149506  1.000580  0.007123   
6       NaN   NaN  0.104519  0.014672  0.006424  0.146442  1.001669  0.008208   
7  0.257569   NaN  0.108926  0.005989  0.000839  0.147615  0.671205  0.001077   
8  0.189361   NaN  0.122928  0.017307  0.004835  0.152756  0.669801  0.006822   
9  0.403448   NaN  0.221899  0.001681  0.002090  0.154326  0.668379  0.007081   

        R_3      D_52       P_3      B_10  D_53       S_5      B_11       S_6  \
0  0.001423  0.207334  0.736463  0.096219   NaN  0.023381  0.002768  0.008322   
1  0.001984  0.202778  0.720886  0.099804   NaN  0.030599  0.002749  0.002482   
2  0.007426  0.206629  0.738044  0.134073   NaN  0.048367  0.010077  0.000530   
3  0.003515  0.208214  0.741813  0.134437   NaN  0.030063  0.009667  0.000783   
4  0.001362  0.205468  0.691986  0.121518   NaN  0.054221  0.009484  0.006698   
5  0.007585  0.20416

In [429]:
#check if any duplicate row
df.duplicated().sum()

0

In [430]:
# Check for missing value
df.isna().sum()

customer_ID         0
S_2                 0
P_2              1542
D_39                0
B_1                 0
B_2                72
R_1                 0
S_3             37665
D_41               72
B_3                72
D_42           170807
D_43            60058
D_44             9929
B_4                 0
D_45               72
B_5                 0
R_2                 0
D_46            43895
D_47                0
D_48            26011
D_49           179538
B_6                 7
B_7                 0
B_8               848
D_50           113536
D_51                0
B_9                 0
R_3                 0
D_52             1009
P_3             10887
B_10                0
D_53           147403
S_5                 0
B_11                0
S_6                 0
D_54               72
R_4                 0
S_7             37665
B_12                0
S_8                 0
D_55             6524
D_56           108881
B_13             1784
R_5                 0
D_58                0
S_9       

---
# Data Preprocessing

---

---
**Handle Variable with Missing Value**

---

---
Drop Variables with Missing Value (>=50%) in train dataset

---

In [431]:
#drop variables with missing values >=50% in the train dataframe
i=0
for col in df.columns:
    if (df[col].isnull().sum()/len(df[col])*100) >=50:
        print("Dropping column", col)
        df.drop(labels=col,axis=1,inplace=True)
        i=i+1
        
print("Total number of columns dropped in train dataframe", i)

Dropping column D_42
Dropping column D_49
Dropping column D_50
Dropping column D_53
Dropping column D_56
Dropping column S_9
Dropping column B_17
Dropping column D_66
Dropping column D_73
Dropping column D_76
Dropping column R_9
Dropping column D_82
Dropping column B_29
Dropping column D_87
Dropping column D_88
Dropping column D_105
Dropping column D_106
Dropping column R_26
Dropping column D_108
Dropping column D_110
Dropping column D_111
Dropping column B_39
Dropping column B_42
Dropping column D_132
Dropping column D_134
Dropping column D_135
Dropping column D_136
Dropping column D_137
Dropping column D_138
Dropping column D_142
Total number of columns dropped in train dataframe 30


In [432]:
df.shape

(200000, 161)

In [433]:
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_columns', 200)

In [434]:
df.isna().sum()

customer_ID        0
S_2                0
P_2             1542
D_39               0
B_1                0
B_2               72
R_1                0
S_3            37665
D_41              72
B_3               72
D_43           60058
D_44            9929
B_4                0
D_45              72
B_5                0
R_2                0
D_46           43895
D_47               0
D_48           26011
B_6                7
B_7                0
B_8              848
D_51               0
B_9                0
R_3                0
D_52            1009
P_3            10887
B_10               0
S_5                0
B_11               0
S_6                0
D_54              72
R_4                0
S_7            37665
B_12               0
S_8                0
D_55            6524
B_13            1784
R_5                0
D_58               0
B_14               0
D_59            3767
D_60               0
D_61           21443
B_15             206
S_11               0
D_62           27433
D_63         

In [435]:
df.head()

customer_ID         S_2       P_2  \
0  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-03-09  0.938469   
1  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-04-07  0.936665   
2  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-05-28  0.954180   
3  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-06-13  0.960384   
4  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-07-16  0.947248   

       D_39       B_1       B_2       R_1       S_3      D_41       B_3  D_43  \
0  0.001733  0.008724  1.006838  0.009228  0.124035  0.008771  0.004709   NaN   
1  0.005775  0.004923  1.000653  0.006151  0.126750  0.000798  0.002714   NaN   
2  0.091505  0.021655  1.009672  0.006815  0.123977  0.007598  0.009423   NaN   
3  0.002455  0.013683  1.002700  0.001373  0.117169  0.000685  0.005531   NaN   
4  0.002483  0.015193  1.000727  0.007605  0.117325  0.004653  0.009312   NaN   

       D_44       B_4      D_45       B_5       R_2      D_46      D_47  \
0  0.000630  0.080986  0.708906  0.170600  0.006204  0.358587  0.525351   
1  0.002526  0.069419  0.712795  0.113239  0.006206  0.353630  0.521311   
2  0.007605  0.068839  0.720884  0.060492  0.003259  0.334650  0.524568   
3  0.006406  0.055630  0.723997  0.166782  0.009918  0.323271  0.530929   
4  0.007731  0.038862  0.720619  0.143630  0.006667  0.231009  0.529305   

       D_48       B_6       B_7       B_8      D_51       B_9       R_3  \
0  0.255736  0.063902  0.059416  0.006466  1.335856  0.008207  0.001423   
1  0.223329  0.065261  0.057744  0.001614  1.339794  0.008373  0.001984   
2  0.189424  0.066982  0.056647  0.005126  1.337179  0.009355  0.007426   
3  0.135586  0.083720  0.049253  0.001418  1.339909  0.006782  0.003515   
4       NaN  0.075900  0.048918  0.001199  1.341735  0.000519  0.001362   

       D_52       P_3      B_10       S_5      B_11       S_6      D_54  \
0  0.207334  0.736463  0.096219  0.023381  0.002768  0.008322  1.001519   
1  0.202778  0.720886  0.099804  0.030599  0.002749  0.002482  1.009033   
2  0.206629  0.738044  0.134073  0.048367  0.010077  0.000530  1.009184   
3  0.208214  0.741813  0.134437  0.030063  0.009667  0.000783  1.007456   
4  0.205468  0.691986  0.121518  0.054221  0.009484  0.006698  1.003738   

        R_4       S_7      B_12       S_8      D_55      B_13       R_5  \
0  0.008298  0.161345  0.148266  0.922998  0.354596  0.118075  0.001882   
1  0.005136  0.140951  0.143530  0.919414  0.326757  0.118737  0.001610   
2  0.006961  0.112229  0.137014  1.001977  0.304124  0.114534  0.006328   
3  0.008706  0.102838  0.129017  0.704016  0.275055  0.120740  0.004980   
4  0.003846  0.094311  0.129539  0.917133  0.231110  0.095178  0.001653   

       D_58      B_14      D_59      D_60      D_61      B_15      S_11  \
0  0.158612  0.018385  0.063646  0.199617  0.308233  0.016361  0.401619   
1  0.148459  0.013035  0.065501  0.151387  0.265026  0.017688  0.406326   
2  0.139504  0.056653  0.070607  0.305883  0.212165  0.063955  0.406768   
3  0.138100  0.012498  0.065926  0.273553  0.204300  0.022732  0.405175   
4  0.126443  0.027897  0.063697  0.233103  0.175655  0.031171  0.487460   

       D_62 D_63 D_64      D_65      B_16      B_18      B_19      B_20  D_68  \
0  0.091071   CR    O  0.007126  0.007665  0.652984  0.008520  0.004730   6.0   
1  0.086805   CR    O  0.002413  0.007148  0.647093  0.002238  0.003879   6.0   
2  0.094001   CR    O  0.001878  0.003636  0.645819  0.000408  0.004578   6.0   
3  0.094854   CR    O  0.005899  0.005896  0.654358  0.005897  0.005207   6.0   
4  0.093915   CR    O  0.009479  0.001714  0.650112  0.007773  0.005851   6.0   

       S_12       R_6      S_13      B_21      D_69      B_22      D_70  \
0  0.272008  0.008363  0.515222  0.002644  0.009013  0.004808  0.008342   
1  0.188970  0.004030  0.509048  0.004193  0.007842  0.001283  0.006524   
2  0.495308  0.006838  0.679257  0.001337  0.006025  0.009393  0.002615   
3  0.508670  0.008183  0.515282  0.008716  

Missing value treatment

In [436]:
#mode imputation on categorical columns
for col in list(df.select_dtypes(['category']).columns):
    df[col].fillna(df[col].mode()[0], inplace=True)

In [437]:
 df[col].dtypes

dtype('int64')

In [438]:
for col in list(df.select_dtypes(['category']).columns):
    df[col] = df[col].astype('Int64')

In [439]:
for col in list(df.select_dtypes(['float64','int32','int64']).columns):

    df[col].fillna(np.mean(~df[col].isnull()), inplace=True)
    #df[col] = df[col].astype(np.float64)

In [440]:
 df[col].dtypes

dtype('int64')

In [441]:
df.reset_index(inplace=True, drop=True)

In [442]:
df.isna().sum()

customer_ID       0
S_2               0
P_2               0
D_39              0
B_1               0
B_2               0
R_1               0
S_3               0
D_41              0
B_3               0
D_43              0
D_44              0
B_4               0
D_45              0
B_5               0
R_2               0
D_46              0
D_47              0
D_48              0
B_6               0
B_7               0
B_8               0
D_51              0
B_9               0
R_3               0
D_52              0
P_3               0
B_10              0
S_5               0
B_11              0
S_6               0
D_54              0
R_4               0
S_7               0
B_12              0
S_8               0
D_55              0
B_13              0
R_5               0
D_58              0
B_14              0
D_59              0
D_60              0
D_61              0
B_15              0
S_11              0
D_62              0
D_63              0
D_64           8040
D_65              0


In [443]:
df.head()

customer_ID         S_2       P_2  \
0  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-03-09  0.938469   
1  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-04-07  0.936665   
2  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-05-28  0.954180   
3  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-06-13  0.960384   
4  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-07-16  0.947248   

       D_39       B_1       B_2       R_1       S_3      D_41       B_3  \
0  0.001733  0.008724  1.006838  0.009228  0.124035  0.008771  0.004709   
1  0.005775  0.004923  1.000653  0.006151  0.126750  0.000798  0.002714   
2  0.091505  0.021655  1.009672  0.006815  0.123977  0.007598  0.009423   
3  0.002455  0.013683  1.002700  0.001373  0.117169  0.000685  0.005531   
4  0.002483  0.015193  1.000727  0.007605  0.117325  0.004653  0.009312   

      D_43      D_44       B_4      D_45       B_5       R_2      D_46  \
0  0.69971  0.000630  0.080986  0.708906  0.170600  0.006204  0.358587   
1  0.69971  0.002526  0.069419  0.712795  0.113239  0.006206  0.353630   
2  0.69971  0.007605  0.068839  0.720884  0.060492  0.003259  0.334650   
3  0.69971  0.006406  0.055630  0.723997  0.166782  0.009918  0.323271   
4  0.69971  0.007731  0.038862  0.720619  0.143630  0.006667  0.231009   

       D_47      D_48       B_6       B_7       B_8      D_51       B_9  \
0  0.525351  0.255736  0.063902  0.059416  0.006466  1.335856  0.008207   
1  0.521311  0.223329  0.065261  0.057744  0.001614  1.339794  0.008373   
2  0.524568  0.189424  0.066982  0.056647  0.005126  1.337179  0.009355   
3  0.530929  0.135586  0.083720  0.049253  0.001418  1.339909  0.006782   
4  0.529305  0.869945  0.075900  0.048918  0.001199  1.341735  0.000519   

        R_3      D_52       P_3      B_10       S_5      B_11       S_6  \
0  0.001423  0.207334  0.736463  0.096219  0.023381  0.002768  0.008322   
1  0.001984  0.202778  0.720886  0.099804  0.030599  0.002749  0.002482   
2  0.007426  0.206629  0.738044  0.134073  0.048367  0.010077  0.000530   
3  0.003515  0.208214  0.741813  0.134437  0.030063  0.009667  0.000783   
4  0.001362  0.205468  0.691986  0.121518  0.054221  0.009484  0.006698   

       D_54       R_4       S_7      B_12       S_8      D_55      B_13  \
0  1.001519  0.008298  0.161345  0.148266  0.922998  0.354596  0.118075   
1  1.009033  0.005136  0.140951  0.143530  0.919414  0.326757  0.118737   
2  1.009184  0.006961  0.112229  0.137014  1.001977  0.304124  0.114534   
3  1.007456  0.008706  0.102838  0.129017  0.704016  0.275055  0.120740   
4  1.003738  0.003846  0.094311  0.129539  0.917133  0.231110  0.095178   

        R_5      D_58      B_14      D_59      D_60      D_61      B_15  \
0  0.001882  0.158612  0.018385  0.063646  0.199617  0.308233  0.016361   
1  0.001610  0.148459  0.013035  0.065501  0.151387  0.265026  0.017688   
2  0.006328  0.139504  0.056653  0.070607  0.305883  0.212165  0.063955   
3  0.004980  0.138100  0.012498  0.065926  0.273553  0.204300  0.022732   
4  0.001653  0.126443  0.027897  0.063697  0.233103  0.175655  0.031171   

       S_11      D_62 D_63 D_64      D_65      B_16      B_18      B_19  \
0  0.401619  0.091071   CR    O  0.007126  0.007665  0.652984  0.008520   
1  0.406326  0.086805   CR    O  0.002413  0.007148  0.647093  0.002238   
2  0.406768  0.094001   CR    O  0.001878  0.003636  0.645819  0.000408   
3  0.405175  0.094854   CR    O  0.005899  0.005896  0.654358  0.005897   
4  0.487460  0.093915   CR    O  0.009479  0.001714  0.650112  0.007773   

       B_20  D_68      S_12       R_6      S_13      B_21      D_69      B_22  \
0  0.004730   6.0  0.272008  0.008363  0.515222  0.002644  0.009013  0.004808   
1  0.003879   6.0  0.188970  0.004030  0.509048  0.004193  0.007842  0.001283   
2  0.004578   6.0  0.495308  0.006838  0.679257  0.001337  0.006025  0.009393   
3  0.005207   6.0  0.508670  0.008183  0.515282  0.008716  0.005271  0.004554   
4  0.005851   6.

In [444]:
df['S_2'] = pd.to_datetime(df['S_2'],format='%Y/%m/%d')

In [445]:
df.dtypes

customer_ID            object
S_2            datetime64[ns]
P_2                   float64
D_39                  float64
B_1                   float64
B_2                   float64
R_1                   float64
S_3                   float64
D_41                  float64
B_3                   float64
D_43                  float64
D_44                  float64
B_4                   float64
D_45                  float64
B_5                   float64
R_2                   float64
D_46                  float64
D_47                  float64
D_48                  float64
B_6                   float64
B_7                   float64
B_8                   float64
D_51                  float64
B_9                   float64
R_3                   float64
D_52                  float64
P_3                   float64
B_10                  float64
S_5                   float64
B_11                  float64
S_6                   float64
D_54                  float64
R_4                   float64
S_7       

In [446]:
from sklearn import preprocessing
  
# label_encoder object knows how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
  
# Encode labels in columns 'D_63' and 'D_64'.
df['D_63']= label_encoder.fit_transform(df['D_63'])
df['D_64']= label_encoder.fit_transform(df['D_64'])  

In [447]:
def find_outliers(col):
    q1 = col.quantile(.25)
    q3 = col.quantile(.75)
    IQR = q3 - q1
    ll = q1 - (1.5*IQR)
    ul = q3 + (1.5*IQR)
    upper_outliers = col[col > ul].index.tolist()
    lower_outliers = col[col < ll].index.tolist()
    bad_indices = list(set(upper_outliers + lower_outliers))
    return(bad_indices)


bad_indexes = []
for col in df.columns:
    if df[col].dtype in ["int64","float64"]:
        bad_indexes.append(find_outliers(df[col]))

bad_indexes = set(list(np.concatenate(bad_indexes).flat))
print(len(bad_indexes))

197216


In [448]:
outliers = []
def detect_outliers_iqr(data):
    data = sorted(data)
    q1 = np.percentile(data, 25)
    q3 = np.percentile(data, 75)
    # print(q1, q3)
    IQR = q3-q1
    lwr_bound = q1-(1.5*IQR)
    upr_bound = q3+(1.5*IQR)
    # print(lwr_bound, upr_bound)
    for i in data: 
        if (i<lwr_bound or i>upr_bound):
            outliers.append(i)
    return outliers

sample_outliers = detect_outliers_iqr(df['P_2'])
print("Outliers from IQR method: ", len(sample_outliers))

Outliers from IQR method:  640


---
# Summary

---

* Variables (Columns) with missing value >= 50% was removed
* OneHotEncoder was used for encoding Categorical Variables
